courtlistener/process_cl.py

```diff
@@ -0,0 +1,132 @@
+import argparse
+import csv
+import logging
+import os
+import re
+import sys
+
+import pandas as pd
+
+from licensed_pile.licenses import PermissiveLicenses
+from licensed_pile.logs import configure_logging
+from licensed_pile.write import to_dolma
+
+SOURCE_NAME = "CourtListenerOpinion"
+
+csv.field_size_limit(sys.maxsize)
+
+logger = configure_logging("court-listener-opinion")
+
+
+def process_court_listener(file_path):
+    df = pd.read_csv(file_path)
+
+    # add metadata column
+    df["metadata"] = str(PermissiveLicenses.PD)
+
+    # add source column
+    df["source"] = SOURCE_NAME
+
+    """
+    Remove columns:
+    date_modified
+    author_str
+    per_curiam
+    joined_by_str
+    type
+    sha1
+    page_count
+    local_path
+    extracted_by_ocr
+    author_id
+    cluster_id
+    """
+    df = df.drop(
+        columns=[
+            "date_modified",
+            "author_str",
+            "per_curiam",
+            "joined_by_str",
+            "type",
+            "sha1",
+            "page_count",
+            "local_path",
+            "extracted_by_ocr",
+            "author_id",
+            "cluster_id",
+        ]
+    )
+
+    """
+    Merge columns based on Court Listener documentation:
+    html_with_citations
+    html_columbia
+    html_lawbox
+    xml_harvard
+    html_anon_2020
+    html
+    plain_text
+    """
+    df["text"] = (
+        df["html_with_citations"]
+        .combine_first(df["html_columbia"])
+        .combine_first(df["html_lawbox"])
+        .combine_first(df["xml_harvard"])
+        .combine_first(df["html_anon_2020"])
+        .combine_first(df["html"])
+    )
+
+    # keep only the text columns and drop null values
+    df = df.drop(
+        columns=[
+            "html",
+            "html_anon_2020",
+            "html_lawbox",
+            "html_columbia",
+            "xml_harvard",
+            "html_with_citations",
+        ]
+    ).dropna(subset=["text"])
+
+    # extract text from html and xml following Harvard CAP
+    # They used r"<.+?>", ""
+    df["text"] = df["text"].apply(lambda x: re.sub(r"<.+?>", "", x))
+
+    # combine merge plain text and extracted text
+    df["text"] = df["text"].combine_first(df["plain_text"])
```

Comment:
This line doesn't do anything, there is already a `.dropna(subset=["text"])` call so there won't be any line where the `text` column is `None` and the `plain_text` column is non-null.

Additionally, `.combine_first` doesn't consider an empty string as null so even if there was a `text` column that was totally removed based on the regex above, it wouldn't get replaced by the `plain_text` column.

The `plain_text` column should be part of the initial `.combine_first` chain or the `.dropna` above should be removed and rows with an empty string for `text` column should be converted to a `None` so the `plain_text` column will actually replace the `text` column

Reply:
This is an artifact of using a different extractor.

Reply:
Trafilatura will over-filter and cause null if not handled explicitly.

Reply:
This is not the case with standard regex.

Reply:
If Trafilatura is so overzelous that it turns real data into null that sounds like a red flag. Also if the `plain_text` column is good enough for those cases is there a reason it isn't good enough to always use it?

Reply:
I agree that it is a red flag. I am writing my own explicit extractor for this dataset but that will not be out until later.

Trafilatura ranks as the highest-quality extractor given a general task.

The extractor should be run on every column which is not plain text. Plain text is considered the lowest quality data in terms of coalescing. It should only be used if the other columns do not exist and can not be extracted.